README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->

Provides simple UI design for chat applications. Developers can focus more on the chat functionality rather than creating unique UI.


## Features

### DemoScreen
<img src="https://raw.githubusercontent.com/DIP-POLLEY/chatbox/main/img_2.png"/>
<br>
<img src="https://raw.githubusercontent.com/DIP-POLLEY/chatbox/main/img_3.png"/>
<br>
<img src="https://raw.githubusercontent.com/DIP-POLLEY/chatbox/main/img_4.png"/>

### Sending Message
<img src="https://raw.githubusercontent.com/DIP-POLLEY/chatbox/main/img.png"/>

### Recieving Message
<img src="https://raw.githubusercontent.com/DIP-POLLEY/chatbox/main/img_1.png"/>


## Getting started

### Requirement
`Dart >=2.17.0 < 3.0.0` and `Flutter >=1.17.0`

## Usage

This is how you can call the chatbox and use the UI.

### Sending
```dart
const ChatBox(
message: "Text message here",
imageURL: "https://picsum.photos/250?image=9",//Sending GIF or Image
recieved: false,
chatBoxColor: Colors.greenAccent,
textColor: Colors.black,
time: "1:15",
),
```

### Receiving
```dart
const ChatBox(
message: "Text message here",
imageURL:"https://docs.flutter.dev/assets/images/dash/dash-fainting.gif",//Receiving GIF or Image
recieved: true,
chatBoxColor: Colors.greenAccent,
textColor: Colors.black,
time: "1:15",
),
```

### Hyperlinks
```dart
const ChatBox(
message: "https://pub.dev/packages/chatbox",
recieved: false,
chatBoxColor: Colors.greenAccent,
textColor: Colors.black,
time: "1:15",
),
```

### Share Documents
```dart
ChatBox(
doculink: "http://www.africau.edu/images/default/sample.pdf",
recieved: false,
chatBoxColor: Colors.greenAccent,
textColor: Colors.black,
time: "1:15",
),
```

## Additional information

You are allowed to raise an issue create a solution and raise a PR. If the issue and the solution are valid it will be merged.

CHANGELOG.md
## 0.0.7

* Adding document sharing support

## 0.0.6

* Adding hyperlink support

## 0.0.5

* Sharing Images support

## 0.0.4

* Fixing documentation images

## 0.0.3

* Updating documentation, adding example

## 0.0.2

* Updating documentation

## 0.0.1

* Initial release 


test/chatbox_test.dart
import 'package:flutter/material.dart';

import 'package:flutter_test/flutter_test.dart';

import 'package:chatbox/chatbox.dart';

void main() {
  test('checks chat box', () {
    const chatbox = ChatBox(
      message: "India is my country",
      recieved: false,
      chatBoxColor: Colors.greenAccent,
      time: "02:15",
    );
    expect(chatbox.message, "India is my country");
    expect(chatbox.recieved, false);
  });
}


lib/src/chatbox.dart
import 'package:any_link_preview/any_link_preview.dart';
import 'package:http/http.dart' as http;

import 'package:flutter/material.dart';
import 'package:url_launcher/url_launcher.dart';

class ChatBox extends StatefulWidget {
  const ChatBox({
    Key? key,
    this.message = "",
    required this.recieved,
    required this.chatBoxColor,
    this.textColor = Colors.black,
    this.imageURL = "",
    this.doculink = "",
    required this.time,
  }) : super(key: key);

  final String message;
  final bool recieved;
  final Color chatBoxColor;
  final Color textColor;
  final String imageURL;
  final String time;
  final String doculink;

  @override
  State<ChatBox> createState() => _ChatBoxState();
}

class _ChatBoxState extends State<ChatBox> {
  bool _recieved = true;
  String _message = "No message was sent";
  Color _chatBoxColor = Colors.greenAccent;
  Color _textColor = Colors.greenAccent;
  String _imageURL = "";
  String time = "";
  String doculink = "";
  String docuname = "";
  bool checknetwork = false;

  final now = DateTime.now();

  void checkconnection() async {
    try {
      final response = await http.get(Uri.parse(_imageURL));
      if (response.statusCode == 200) {
        setState(() {
          checknetwork = true;
        });
      } else {
        setState(() {
          checknetwork = false;
        });
      }
    } catch (e) {
      //print(e);
    }
  }

  void verifyconditions() {
    if (_imageURL == "" && _message == "") {
      setState(() {
        _message = "Nothing to display";
      });
    } else if (_message == "") {
      setState(() {
        _message = " ";
      });
    }
  }

  @override
  void initState() {
    super.initState();
    _recieved = widget.recieved;
    _message = widget.message;
    _chatBoxColor = widget.chatBoxColor;
    _textColor = widget.textColor;
    _imageURL = widget.imageURL;
    time = widget.time;
    doculink = widget.doculink;
    if (_imageURL != "") {
      checkconnection();
    }
    if (doculink != "") {
      getname();
    }
    verifyconditions();
  }

  void getname() {
    var filename = doculink.split('/').last;
    String s = "";
    for (int i = 0; i < filename.length; i++) {
      if (filename[i] == "?") {
        break;
      }
      s = s + filename[i];
    }
    setState(() {
      docuname = s;
    });
    // print(s);
  }

  Future openlink() async {
    await launchUrl(Uri.parse(doculink));
    //final taskid = await Flu
  }

  @override
  Widget build(BuildContext context) {
    return _recieved == true
        ? Align(
            alignment: Alignment.topLeft,
            child: Padding(
              padding: const EdgeInsets.only(top: 10, left: 10),
              child: Container(
                // width: 250,
                constraints: BoxConstraints(
                  maxWidth: MediaQuery.of(context).size.width * 2 / 3,
                  minWidth: MediaQuery.of(context).size.width / 4,
                  minHeight: MediaQuery.of(context).size.height / 15,
                ),
                decoration: BoxDecoration(
                  color: _chatBoxColor,
                  borderRadius: const BorderRadius.only(
                    topRight: Radius.circular(10),
                    bottomLeft: Radius.circular(10),
                    bottomRight: Radius.circular(10),
                  ),
                ),
                child: doculink != ""
                    ? Padding(
                        padding: const EdgeInsets.all(5.0),
                        child: Column(
                          mainAxisAlignment: MainAxisAlignment.start,
                          children: [
                            Container(
                              decoration: BoxDecoration(
                                color: Colors.black45.withOpacity(0.1),
                                borderRadius: const BorderRadius.only(
                                  topRight: Radius.circular(10),
                                  bottomLeft: Radius.circular(10),
                                  bottomRight: Radius.circular(10),
                                ),
                              ),
                              child: Row(
                                children: [
                                  const Expanded(
                                    flex: 1,
                                    child: Padding(
                                      padding: EdgeInsets.all(10),
                                      child: Icon(
                                        Icons.insert_drive_file_rounded,
                                      ),
                                    ),
                                  ),
                                  Expanded(
                                    flex: 9,
                                    child: Padding(
                                      padding: const EdgeInsets.all(10),
                                      child: Text(
                                        docuname,
                                        style: const TextStyle(fontSize: 15),
                                      ),
                                    ),
                                  )
                                ],
                              ),
                            ),
                            Align(
                              alignment: Alignment.topLeft,
                              child: Text(
                                time,
                                textAlign: TextAlign.left,
                              ),
                            )
                          ],
                        ),
                      )
                    : Padding(
                        padding: const EdgeInsets.all(5.0),
                        child: _imageURL != ""
                            ? Column(
                                mainAxisSize: MainAxisSize.min,
                                crossAxisAlignment: CrossAxisAlignment.start,
                                children: [
                                  ClipRRect(
                                    borderRadius: BorderRadius.circular(5),
                                    child: !checknetwork
                                        ? Image.asset(
                                            "packages/chatbox/assets/loading.gif",
                                            height: 10,
                                          )
                                        : Image.network(_imageURL,
                                            loadingBuilder:
                                                (BuildContext context,
                                                    Widget child,
                                                    ImageChunkEvent?
                                                        loadingProgress) {
                                            if (loadingProgress == null) {
                                              return child;
                                            }
                                            return SizedBox(
                                              height: 100,
                                              child: Center(
                                                child:
                                                    CircularProgressIndicator(
                                                  value: loadingProgress
                                                              .expectedTotalBytes !=
                                                          null
                                                      ? loadingProgress
                                                              .cumulativeBytesLoaded /
                                                          loadingProgress
                                                              .expectedTotalBytes!
                                                      : null,
                                                ),
                                              ),
                                            );
                                          }),
                                  ),
                                  TextBox(
                                    message: _message,
                                    textColor: _textColor,
                                    time: time,
                                  )
                                ],
                              )
                            : TextBox(
                                message: _message,
                                textColor: _textColor,
                                time: time,
                              )),
              ),
            ),
          )
        : Align(
            alignment: Alignment.topRight,
            child: Padding(
              padding: const EdgeInsets.only(top: 10, left: 120, right: 10),
              child: Container(
                constraints: BoxConstraints(
                  maxWidth: MediaQuery.of(context).size.width * 2 / 3,
                  minWidth: MediaQuery.of(context).size.width / 5,
                  minHeight: MediaQuery.of(context).size.height / 15,
                ),
                decoration: BoxDecoration(
                  color: _chatBoxColor,
                  borderRadius: const BorderRadius.only(
                    topLeft: Radius.circular(10),
                    bottomLeft: Radius.circular(10),
                    bottomRight: Radius.circular(10),
                  ),
                ),
                child: doculink != ""
                    ? Padding(
                        padding: const EdgeInsets.all(5.0),
                        child: Column(
                          mainAxisAlignment: MainAxisAlignment.start,
                          children: [
                            GestureDetector(
                              onTap: () {
                                openlink();
                              },
                              child: Container(
                                decoration: BoxDecoration(
                                  color: Colors.black45.withOpacity(0.1),
                                  borderRadius: const BorderRadius.only(
                                    topLeft: Radius.circular(10),
                                    bottomLeft: Radius.circular(10),
                                    bottomRight: Radius.circular(10),
                                  ),
                                ),
                                child: Row(
                                  children: [
                                    const Expanded(
                                      flex: 1,
                                      child: Padding(
                                        padding: EdgeInsets.all(10),
                                        child: Icon(
                                          Icons.insert_drive_file_rounded,
                                        ),
                                      ),
                                    ),
                                    Expanded(
                                      flex: 9,
                                      child: Padding(
                                        padding: const EdgeInsets.all(10),
                                        child: Text(
                                          docuname,
                                          style: const TextStyle(fontSize: 15),
                                        ),
                                      ),
                                    )
                                  ],
                                ),
                              ),
                            ),
                            Align(
                              alignment: Alignment.topRight,
                              child: Text(
                                time,
                                textAlign: TextAlign.right,
                              ),
                            )
                          ],
                        ),
                      )
                    : Padding(
                        padding: const EdgeInsets.all(5.0),
                        child: _imageURL != ""
                            ? Column(
                                mainAxisSize: MainAxisSize.min,
                                crossAxisAlignment: CrossAxisAlignment.start,
                                children: [
                                  ClipRRect(
                                    borderRadius: BorderRadius.circular(5),
                                    child: !checknetwork
                                        ? Image.asset(
                                            "packages/chatbox/assets/loading.gif",
                                            height: 10,
                                          )
                                        : Image.network(_imageURL,
                                            loadingBuilder:
                                                (BuildContext context,
                                                    Widget child,
                                                    ImageChunkEvent?
                                                        loadingProgress) {
                                            if (loadingProgress == null) {
                                              return child;
                                            }
                                            return SizedBox(
                                              height: 100,
                                              child: Center(
                                                child:
                                                    CircularProgressIndicator(
                                                  value: loadingProgress
                                                              .expectedTotalBytes !=
                                                          null
                                                      ? loadingProgress
                                                              .cumulativeBytesLoaded /
                                                          loadingProgress
                                                              .expectedTotalBytes!
                                                      : null,
                                                ),
                                              ),
                                            );
                                          }),
                                  ),
                                  TextBox(
                                    message: _message,
                                    textColor: _textColor,
                                    time: time,
                                  )
                                ],
                              )
                            : TextBox(
                                message: _message,
                                textColor: _textColor,
                                time: time,
                              )),
              ),
            ),
          );
  }
}

class TextBox extends StatefulWidget {
  const TextBox({
    Key? key,
    required String message,
    required Color textColor,
    required this.time,
  })  : _message = message,
        _textColor = textColor,
        super(key: key);

  final String _message;
  final Color _textColor;
  final String time;

  @override
  State<TextBox> createState() => _TextBoxState();
}

class _TextBoxState extends State<TextBox> {
  String url = "";

  @override
  Widget build(BuildContext context) {
    final matcher = RegExp(
        r"(http(s)?:\/\/.)?(www\.)?[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)");

    if (matcher.stringMatch(widget._message) != null) {
      setState(() {
        url = matcher.stringMatch(widget._message).toString();
      });
    }

    return url != ""
        ? Column(
            mainAxisSize: MainAxisSize.min,
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              AnyLinkPreview(
                link: url,
                cache: const Duration(hours: 1),
                displayDirection: UIDirection.uiDirectionHorizontal,
                backgroundColor: Colors.white,
                onTap: () {
                  launchUrl(Uri.parse(url));
                },
              ),
              RichText(
                text: TextSpan(
                  text: widget._message != "" ? "${widget._message}\n" : "\n",
                  style: TextStyle(
                    color: widget._textColor,
                    fontWeight: FontWeight.bold,
                  ),
                  children: <TextSpan>[
                    TextSpan(
                        text: widget.time,
                        //"${(widget.now.hour).toString().padLeft(2, "0")}:${(widget.now.minute).toString().padLeft(2, "0")}",
                        style: const TextStyle(
                          color: Colors.blueGrey,
                          fontWeight: FontWeight.w500,
                        )),
                  ],
                ),
              ),
            ],
          )
        : RichText(
            text: TextSpan(
              text: widget._message != "" ? "${widget._message}\n" : "\n",
              style: TextStyle(
                color: widget._textColor,
                fontWeight: FontWeight.bold,
              ),
              children: <TextSpan>[
                TextSpan(
                    text: widget.time,
                    //"${(widget.now.hour).toString().padLeft(2, "0")}:${(widget.now.minute).toString().padLeft(2, "0")}",
                    style: const TextStyle(
                      color: Colors.blueGrey,
                      fontWeight: FontWeight.w500,
                    )),
              ],
            ),
          );
  }
}


example/lib/main.dart
import 'package:chatbox/chatbox.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return SafeArea(
        child: Scaffold(
      body: ListView(
        children: const [
          ChatBox(
            message: "Hi",
            recieved: false,
            chatBoxColor: Colors.greenAccent,
            textColor: Colors.black,
            time: "1:15",
          ),
          ChatBox(
            message: "Hello",
            recieved: true,
            chatBoxColor: Colors.greenAccent,
            textColor: Colors.black,
            time: "1:15",
          ),
          ChatBox(
            message: "Now we can share images also",
            imageURL: "https://picsum.photos/250?image=9",
            recieved: false,
            chatBoxColor: Colors.greenAccent,
            textColor: Colors.black,
            time: "1:15",
          ),
          ChatBox(
            message: "Yes Gifs are supported now",
            imageURL:
                "https://docs.flutter.dev/assets/images/dash/dash-fainting.gif",
            recieved: true,
            chatBoxColor: Colors.greenAccent,
            textColor: Colors.black,
            time: "1:15",
          ),
          ChatBox(
            message: "https://pub.dev/packages/chatbox",
            recieved: false,
            chatBoxColor: Colors.greenAccent,
            textColor: Colors.black,
            time: "1:15",
          ),
          ChatBox(
            doculink:
                "https://firebasestorage.googleapis.com/v0/b/whatsapp-4a91a.appspot.com/o/loading.gif?alt=media&token=268c0d30-f552-4fb0-800d-0cb5506e24c5",
            recieved: true,
            chatBoxColor: Colors.greenAccent,
            textColor: Colors.black,
            time: "1:15",
          ),
          ChatBox(
            doculink: "http://www.africau.edu/images/default/sample.pdf",
            recieved: false,
            chatBoxColor: Colors.greenAccent,
            textColor: Colors.black,
            time: "1:15",
          ),
          SizedBox(
            height: 30,
          )
        ],
      ),
    ));
  }
}
